cmd/clusterctl/cmd/init.go

```diff
@@ -0,0 +1,96 @@
+/*
+Copyright 2019 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package cmd
+
+import (
+	"github.com/spf13/cobra"
+)
+
+type initOptions struct {
+	kubeconfig              string
+	coreProvider            string
+	bootstrapProviders      []string
+	infrastructureProviders []string
+	targetNamespace         string
+	watchingNamespace       string
+	force                   bool
+}
+
+var io = &initOptions{}
+
+var initCmd = &cobra.Command{
+	Use:   "init",
+	Short: "Initialize a management cluster for Cluster API",
+	Long: LongDesc(`
+		Initialize a management cluster for Cluster API by installing cluster API core components,
+		the kubeadm bootstrap provider, and the selected bootstrap and infrastructure providers.
+
+		The management cluster should be an existing Kubernetes cluster, and the identity used
+		for accessing the cluster must have enough privileges for installing cluster API providers.
+
+		Use 'clusterctl config providers' to get the list of available providers; if necessary, edit
+		the .clusterctl file to add new provider configurations or to customize existing ones.
+		
+		Some providers require environment variable to be set before running clusterctl init; please
+		refer to the provider documentation or use 'clusterctl config provider [name]' to get the
+		list of required variables.
+
+		See https://cluster-api.sigs.k8s.io/ for more details.`),
+
+	Example: Examples(`
+		# Initialize a management cluster, the cluster currently pointed by kubectl, by installing the
+		# AWS infrastructure provider. Please note that this command, when executed on an empty management cluster, 
+ 		# automatically triggers the installation of the cluster-api core provider.
+		clusterctl init --infrastructure=aws
```

Comment:
A bit unrelated to this actual PR, but I am wondering a bit about the implementation. How are you imagine this is implemented. Will the CLI be directly initializing the AWS cluster, or will it use a local bootstrap cluster and pivot from there to the AWS cluster?

Reply:
clusterctl v2 behaves differently from clusterctlv1,
init now will be responsible to initialize a management cluster provided by the user, and in this example, to installs the AWS infra provider in the management cluster. No action will be taken by clusterctl on AWS (or on any other cloud provider).

Might be reading the [CAEP](https://github.com/kubernetes-sigs/cluster-api/blob/master/docs/proposals/20191016-clusterctl-redesign.md) can provide more context; feel free to reach out to me for any further question.

Reply:
I read the CAEP briefly in a very early stage, so I probably overlooked that design choice. Thanks for the context @fabriziopandini. 🙂